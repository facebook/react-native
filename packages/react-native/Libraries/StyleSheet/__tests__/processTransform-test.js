/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @oncall react_native
 */

'use strict';

const processTransform = require('../processTransform');

describe('processTransform', () => {
  describe('validation', () => {
    it('should accept an empty array', () => {
      processTransform([]);
    });

    it('should accept an empty string', () => {
      processTransform('');
    });

    it('should accept a simple valid transform', () => {
      processTransform([
        {scale: 0.5},
        {translateX: 10},
        {translateY: 20},
        {rotate: '10deg'},
      ]);
      processTransform(
        'scale(0.5) translateX(10px) translateY(20px) rotate(10deg)',
      );
    });

    it('should accept a percentage translate transform', () => {
      processTransform([{translateY: '20%'}, {translateX: '10%'}]);
      processTransform('translateX(10%)');
    });

    it('should throw on object with multiple properties', () => {
      expect(() =>
        processTransform([{scale: 0.5, translateY: 10}]),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw on invalid transform property', () => {
      expect(() =>
        processTransform([{translateW: 10}]),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        processTransform('translateW(10)'),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw when not passing an array to an array prop', () => {
      expect(() =>
        processTransform([{matrix: 'not-a-matrix'}]),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        processTransform([{translate: 10}]),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should accept a valid matrix', () => {
      processTransform([{matrix: [1, 1, 1, 1, 1, 1, 1, 1, 1]}]);
      processTransform('matrix(1, 1, 1, 1, 1, 1, 1, 1, 1)');
      processTransform([
        {matrix: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]},
      ]);
      processTransform(
        'matrix(1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1)',
      );
    });

    it('should throw when passing a matrix of the wrong size', () => {
      expect(() =>
        processTransform([{matrix: [1, 1, 1, 1]}]),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        processTransform('matrix(1, 1, 1, 1)'),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should accept a valid translate', () => {
      processTransform([{translate: [1, 1]}]);
      processTransform('translate(1px)');
      processTransform('translate(1px, 1px)');
      processTransform([{translate: [1, 1, 1]}]);
    });

    it('should throw when passing a translate of the wrong size', () => {
      expect(() =>
        processTransform([{translate: [1]}]),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        processTransform([{translate: [1, 1, 1, 1]}]),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        processTransform('translate(1px, 1px, 1px, 1px)'),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw when passing an invalid value to a number prop', () => {
      expect(() =>
        processTransform([{translateY: '20deg'}]),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        processTransform([{scale: {x: 10, y: 10}}]),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        processTransform([{perspective: []}]),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw when passing a perspective of 0', () => {
      expect(() =>
        processTransform([{perspective: 0}]),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should accept an angle in degrees or radians', () => {
      processTransform([{skewY: '10deg'}]);
      processTransform('skewY(10deg)');
      processTransform([{rotateX: '1.16rad'}]);
      processTransform('rotateX(1.16rad)');
    });

    it('should throw when passing an invalid angle prop', () => {
      expect(() =>
        processTransform([{rotate: 10}]),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        processTransform('rotate(10)'),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        processTransform([{skewX: '10drg'}]),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        processTransform('skewX(10drg)'),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should throw when passing an Animated.Value', () => {
      expect(() =>
        processTransform([{rotate: {getValue: () => {}}}]),
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
