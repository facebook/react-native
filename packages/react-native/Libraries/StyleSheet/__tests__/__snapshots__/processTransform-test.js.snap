// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processTransform validation should throw on invalid transform property 1`] = `"Invalid transform translateW: {\\"translateW\\":10}"`;

exports[`processTransform validation should throw on invalid transform property 2`] = `"Invalid transform translateW: {\\"translateW\\":10}"`;

exports[`processTransform validation should throw on object with multiple properties 1`] = `"You must specify exactly one property per transform object. Passed properties: {\\"scale\\":0.5,\\"translateY\\":10}"`;

exports[`processTransform validation should throw when not passing an array to an array prop 1`] = `"Transform with key of matrix must have an array as the value: {\\"matrix\\":\\"not-a-matrix\\"}"`;

exports[`processTransform validation should throw when not passing an array to an array prop 2`] = `"Transform with key of translate must have an array as the value: {\\"translate\\":10}"`;

exports[`processTransform validation should throw when passing a matrix of the wrong size 1`] = `"Matrix transform must have a length of 9 (2d) or 16 (3d). Provided matrix has a length of 4: {\\"matrix\\":[1,1,1,1]}"`;

exports[`processTransform validation should throw when passing a matrix of the wrong size 2`] = `"Matrix transform must have a length of 9 (2d) or 16 (3d). Provided matrix has a length of 4: {\\"matrix\\":[1,1,1,1]}"`;

exports[`processTransform validation should throw when passing a perspective of 0 1`] = `"Transform with key of \\"perspective\\" cannot be zero: {\\"perspective\\":0}"`;

exports[`processTransform validation should throw when passing a translate of the wrong size 1`] = `"Transform with key translate must be an array of length 2 or 3, found 1: {\\"translate\\":[1]}"`;

exports[`processTransform validation should throw when passing a translate of the wrong size 2`] = `"Transform with key translate must be an array of length 2 or 3, found 4: {\\"translate\\":[1,1,1,1]}"`;

exports[`processTransform validation should throw when passing a translate of the wrong size 3`] = `"Transform with key translate must be an string with 1 or 2 parameters, found 4: translate(1px, 1px, 1px, 1px)"`;

exports[`processTransform validation should throw when passing an Animated.Value 1`] = `"You passed an Animated.Value to a normal component. You need to wrap that component in an Animated. For example, replace <View /> by <Animated.View />."`;

exports[`processTransform validation should throw when passing an invalid angle prop 1`] = `"Transform with key of \\"rotate\\" must be a string: {\\"rotate\\":10}"`;

exports[`processTransform validation should throw when passing an invalid angle prop 2`] = `"Transform with key of \\"rotate\\" must be a string: {\\"rotate\\":10}"`;

exports[`processTransform validation should throw when passing an invalid angle prop 3`] = `"Rotate transform must be expressed in degrees (deg) or radians (rad): {\\"skewX\\":\\"10drg\\"}"`;

exports[`processTransform validation should throw when passing an invalid angle prop 4`] = `"Rotate transform must be expressed in degrees (deg) or radians (rad): {\\"skewX\\":\\"10drg\\"}"`;

exports[`processTransform validation should throw when passing an invalid value to a number prop 1`] = `"Transform with key of \\"translateY\\" must be number or a percentage. Passed value: {\\"translateY\\":\\"20deg\\"}."`;

exports[`processTransform validation should throw when passing an invalid value to a number prop 2`] = `"Transform with key of \\"scale\\" must be a number: {\\"scale\\":{\\"x\\":10,\\"y\\":10}}"`;

exports[`processTransform validation should throw when passing an invalid value to a number prop 3`] = `"Transform with key of \\"perspective\\" must be a number: {\\"perspective\\":[]}"`;
